src/coreclr/runtime/i386/AllocFast.S

```diff
@@ -0,0 +1,275 @@
+
+        jmp         RhExceptionHandling_FailedAllocation
+LEAF_END RhpNewObject, _TEXT
+
+// Shared code for RhNewString, RhpNewArrayFast and RhpNewObjectArray
+//  EAX == string/array size
+//  ECX == MethodTable
+//  EDX == character/element count
+.macro NEW_ARRAY_FAST_PROLOG
+        ESP_PROLOG_BEG
+        ESP_PROLOG_PUSH ecx
+        ESP_PROLOG_PUSH edx
+        ESP_EPILOG_END
+.endm
+
+.macro NEW_ARRAY_FAST
+        // edx = ee_alloc_context pointer, TRASHES ecx
+        INLINE_GET_ALLOC_CONTEXT
+
+        // ECX == scratch
+        // EAX == allocation size
+        // EDX == ee_alloc_context pointer
+
+        mov         ecx, eax
+        add         eax, [edx + OFFSETOF__ee_alloc_context__alloc_ptr]
+        jc          1f
+        cmp         eax, [edx + OFFSETOF__ee_alloc_context__combined_limit]
+        ja          1f
+
+        // ECX == allocation size
+        // EAX == new alloc ptr
+        // EDX == ee_alloc_context pointer
+
+        // set the new alloc pointer
+        mov         [edx + OFFSETOF__ee_alloc_context__alloc_ptr], eax
+
+        // calc the new object pointer
+        sub         eax, ecx
+
+        ESP_EPILOG_BEG
+        // Restore the element count and put it in edx
+        ESP_EPILOG_POP edx
+        // Restore the MethodTable and put it in ecx
+        ESP_EPILOG_POP ecx
+        ESP_EPILOG_END
+
+        // set the new object's MethodTable pointer and element count
+        mov         [eax + OFFSETOF__Object__m_pEEType], ecx
+        mov         [eax + OFFSETOF__Array__m_Length], edx
+        ret
+
+1:
+        ESP_EPILOG_BEG
+        // Restore the element count and put it in edx
+        ESP_EPILOG_POP edx
+        // Restore the MethodTable and put it in ecx
+        ESP_EPILOG_POP ecx
+        ESP_EPILOG_END
+
+        jmp         RhpNewArray
+.endm
+
+// Allocate a new string.
+//  ECX == MethodTable
+//  EDX == element count
+LEAF_ENTRY RhNewString, _TEXT
+        // Make sure computing the aligned overall allocation size won't overflow
+        cmp         edx, MAX_STRING_LENGTH
+        ja          LOCAL_LABEL(RhNewString_StringSizeOverflow)
+
+        // Compute overall allocation size (align(base size + (element size * elements), 4)).
+        lea         eax, [(edx * STRING_COMPONENT_SIZE) + (STRING_BASE_SIZE + 3)]
+        and         eax, -4
+
+        NEW_ARRAY_FAST_PROLOG
+        NEW_ARRAY_FAST
+
+LOCAL_LABEL(RhNewString_StringSizeOverflow):
+        // We get here if the size of the final string object can't be represented as an unsigned
+        // 32-bit value. We're going to tail-call to a managed helper that will throw
+        // an OOM exception that the caller of this allocator understands.
+
+        // ecx holds MethodTable pointer already
+        xor         edx, edx            // Indicate that we should throw OOM.
+        jmp         RhExceptionHandling_FailedAllocation
+LEAF_END RhNewString, _TEXT
+
+
+// Allocate one dimensional, zero based array (SZARRAY).
```

Comment:
Something is off here. Either the comment is wrong or there is actual bug somewhere.

This is path is hit by allocations of multi-dimensional arrays on NAOT - stacktrace for `new int[1,1]`:
```
repro!RhpNewArrayFast [C:\runtime\src\coreclr\runtime\amd64\AllocFast.asm @ 156] 
repro!S_P_CoreLib_System_Array__NewMultiDimArray+0x5b [C:\runtime\src\coreclr\nativeaot\System.Private.CoreLib\src\System\Array.NativeAot.cs @ 669] 
repro!repro_Program___Main__+0x2a [C:\repro\Program.cs @ 1] 
```

Reply:
I see the `NewMultiDimArray` -> `RhNewArray` -> `RhpNewArrayFast` code path which would make the comment wrong. The comment is, however, just copied from the other pre-existing `RhpNewArray` (before rename) implementations.

The intention seems to be to allow MD arrays with `RhpNewArrayFast` with the size being set to the total array size. I'm just guessing on reading the pre-existing code though.

Reply:
Yes, it appears to be the intention, but it does not work correctly for boundary cases:

`Console.WriteLine(new int[0x70000000,2].LongLength);` works on regular CoreCLR, fails on NAOT with OOM.

Reply:
Sacrificing the fast path for MD arrays sounds reasonable. Regular CoreCLR does not have fast path for MD arrays either.